shared.js

```diff
@@ -217,6 +228,40 @@ class MemFS {
     return ESUCCESS;
   }
 
+  host_read(fd, iovs, iovs_len, nread) {
+    this.hostMem_.check();
+    assert(fd === 0);
+    let size = 0;
+    for (let i = 0; i < iovs_len; ++i) {
+      const data = this.hostMem_.read32(iovs);
+      iovs += 4;
+      const len = this.hostMem_.read32(iovs);
+      iovs += 4;
+      const lenToWrite = len < (this.stdinStr.length - this.stdinStrPos) ? len : (this.stdinStr.length - this.stdinStrPos);
+      if(lenToWrite === 0){
+        break;
+      }
+      const buf = this.hostMem_.read(data,len)
+      // copy the original data out
+      const strToWrite = this.stdinStr.substr(0,lenToWrite).split('').map(x => x.charCodeAt(0));
+      // generate the substr to write
+      const subBuf = new Uint8Array(buf,0,lenToWrite)
+      subBuf.set(strToWrite)
+      // write the substr to buf
+      this.hostMem_.write(data,buf)
+      // write buf back
+      size += lenToWrite;
+      this.stdinStrPos += lenToWrite;
+      if(lenToWrite !== len){
```

Comment:
I'm not sure if this is correct behavior; I think the rest of the reads should still go through, even if the previous one fails.

Reply:
when the `lenToWrite` is not equivalent with the `len`, it means that there is nothing more to read, we reached the EOF of the stdin file. I wrote `break` to speed things up, but when reading from regular files as you can see [here](https://github.com/binji/llvm-project/blob/master/binji/memfs.c#L575), it will return some random data when reached EOF? I'm not sure about this.

Reply:
Checked the `memfs.c` again, I still can't figure out how the procedure of reading from a regular file works. [These lines](https://github.com/binji/llvm-project/blob/master/binji/memfs.c#L572-L574) seems will write the whole file to the buf at once even the buf is smaller than the file? Have you tested this on use cases like reading large files?

Reply:
No you're right. These functions are meant to work like linux `readv`. The confusion here is the difference between `readv` and `preadv` I think. `ReadIovec` assumes that the offset is fixed (like `preadv`), so it won't work properly if there are multiple buffers on a `readv` call.

Reply:
So should we break here? In my opinion, we should, because we reached the end of the input string, continue reading seems useless.

Reply:
Yes, I think so, at least for now. Ultimately I think we may want a separate call, so `preadv` and `readv` can be handled separately when interacting with JS. But if most applications are treating `stdin` as stream-like, then updating the offset and stopping when there's no more input seems like the right choice.